lucene/core/src/java/org/apache/lucene/index/PooledDocValuesReader.java

```diff
@@ -0,0 +1,548 @@
+    if (fieldFilter.test(field) == false) {
+      return super.getSortedSetDocValues(field);
+    }
+    if (pooledValues.containsKey(field) == false) {
+      SortedSetDocValues ssdv = in.getSortedSetDocValues(field);
+      if (ssdv == null) {
+        return null;
+      }
+      pooledValues.put(field, new PooledSortedSet(ssdv));
+    }
+    PooledIterator it = pooledValues.get(field);
+    if (it instanceof PooledSortedSet == false) {
+      return null;
+    }
+    return ((PooledSortedSet)it).getIterator();
+  }
+
+  @Override
+  public CacheHelper getCoreCacheHelper() {
+    return in.getCoreCacheHelper();
+  }
+
+  @Override
+  public CacheHelper getReaderCacheHelper() {
+    return in.getReaderCacheHelper();
+  }
+
+  private interface PooledIterator {
+    void advanceReal(int target) throws IOException;
+  }
+
+  private class PooledNumeric extends NumericDocValues implements PooledIterator {
+
+    final NumericDocValues in;
+    boolean positioned = false;
+    int realDoc = -1;
+
+    private PooledNumeric(NumericDocValues in) {
+      this.in = in;
+    }
+
+    @Override
+    public void advanceReal(int target) throws IOException {
+      if (target > realDoc) {
+        realDoc = in.advance(target);
+      }
+      positioned = realDoc == target;
+    }
+
+    @Override
+    public long longValue() throws IOException {
+      return in.longValue();
+    }
+
+    @Override
+    public boolean advanceExact(int target) throws IOException {
+      if (target != currentDoc) {
+        throw new IllegalStateException("Pooled docvalues can only be advanced to their controlling doc " + target);
+      }
+      return positioned;
+    }
+
+    @Override
+    public int docID() {
+      return in.docID();
+    }
+
+    @Override
+    public int nextDoc() throws IOException {
+      throw new IllegalStateException("Cannot call nextDoc() on pooled docvalues");
+    }
+
+    @Override
+    public int advance(int target) throws IOException {
+      if (target != currentDoc) {
+        throw new IllegalStateException("Pooled docvalues can only be advanced to their controlling doc " + target);
+      }
+      return realDoc;
+    }
+
+    @Override
+    public long cost() {
+      return in.cost();
+    }
+  }
+
+  private class PooledSortedNumeric implements PooledIterator {
+
+    final SortedNumericDocValues in;
+    boolean positioned;
+    int realDoc = -1;
+    long[] values = new long[1];
+
+    private PooledSortedNumeric(SortedNumericDocValues in) {
+      this.in = in;
+    }
+
+    @Override
+    public void advanceReal(int target) throws IOException {
+      if (target > realDoc) {
+        realDoc = in.advance(target);
+      }
+      positioned = realDoc == target;
+      if (positioned) {
+        int c = in.docValueCount();
+        values = ArrayUtil.grow(values, c);
+        for (int i = 0; i < c; i++) {
+          values[i] = in.nextValue();
```

Comment:
I don't understand why it's necessary to cache these per doc values. DocValues seems naturally sharable across `SortingGroupHeads` every one of them calls `advanceExact()` which is valid for the same doc that allow to get through the same values again and again.